docs: fix due to change in hardhat-templates + README (#642)
zama-ai/fhevm: docs/fundamentals/asXXoperation.md

> # asEbool, asUintXX and asBytesXX operations

Reviewer: Wdyt about renaming the folder  `asE[type]operator.md`?

> # asEbool, asUintXX and asEbytesXX operations

Reviewer: same here, replace asUintXX with asEuintXX, and add asEaddress

> This documentation covers the `asEbool`, `asEuintXX`, and `asBytesXX` operations provided by the TFHE library for working with encrypted data in the fhEVM. These operations are essential for converting between plaintext and encrypted types, as well as handling encrypted inputs.

Reviewer: should be asEbool, asEuintXX and as EbBtesXX, both asEuintXX and as EBytesXX are wrong in the title, and asEbytes in the text as well

The operations can be categorized into three main use cases:

1. **Trivial encryption**: Converting plaintext values to encrypted types
2. **Type casting**: Converting between different encrypted types
3. **Input handling**: Processing encrypted inputs with proofs

> ## 1. Trivial input

Reviewer: Should be "Trivial Encryption" everywhere, not trivial input

Author: we had a discussion with Levent on this it's a bit easier to clasify

Author: or you know what yeah will just convert to trivial encrypt

Reviewer: as you wish, but I think Trivial encryption makes more sense because few lines earlier you already used trivial encryption, so let's not use different terminologies for same concept

> Trivial input simply put is a plain text in a format of a cyphertext.

Reviewer 2: ```suggestion
Trivial input simply puts plaintext into ciphertext format.
```

### Overview

Trivial input is the process of converting plaintext values into encrypted types (ciphertexts) compatible with TFHE operators. Although the data is in ciphertext format, it remains publicly visible on-chain, making it useful for operations between public and private values.

This type of casting involves converting plaintext (unencrypted) values into their encrypted equivalents, such as:

- `bool` → `ebool`
- `uint` → `euintXX`
- `bytes` → `ebytesXX`

> > **Note**: When doing trivial encrypt, the data is made compatible with FHE operations but remains publicly visible on-chain unless explicitly encrypted.

Reviewer: trivial encrypt*ion*

> > **Note**: When doing trivial encrypt, the data is made compatible with FHE operations but remains publicly visible on-chain unless explicitly encrypted.

Author: ```suggestion
> **Note**: When doing trivial encryption, the data is made compatible with FHE operations but remains publicly visible on-chain unless explicitly encrypted.
```

#### **Example**

```solidity
euint64 value64 = TFHE.asEuint64(7262);  // Trivial encrypt a uint64
ebool valueBool = TFHE.asEbool(true);   // Trivial encrypt a boolean
```

### Trivial encryption of `ebytesXX` types

The `TFHE.padToBytesXX` functions facilitate this trivial encryption process for byte arrays, ensuring compatibility with `ebytesXX` types. These functions:

- Pad the provided byte array to the appropriate length (`64`, `128`, or `256` bytes).
- Prevent runtime errors caused by improperly sized input data.
- Work seamlessly with `TFHE.asEbytesXX` for trivial encryption.

> **Important**: Trivial encryption does NOT provide any privacy guarantees. The input data remains fully visible on the blockchain. Only use trivial encryption when working with public values that need to interact with actual encrypted data.

#### Workflow

1. **Pad Input Data**:
   Use the `padToBytesXX` functions to ensure your byte array matches the size requirements.
2. **Encrypt the Padded Data**:
   Use `TFHE.asEbytesXX` to encrypt the padded byte array into the corresponding encrypted type.
3. **Grant Access**:

>    Use `TFHE.allowThis` and `TFHE.allow` to define access control for the encrypted data.

Reviewer: "3. Optionally, if you want to persist allowance for those variables for later use, ..."

>    Use `TFHE.allowThis` and `TFHE.allow` to define access control for the encrypted data.

Author: ```suggestion
   Use `TFHE.allowThis` and `TFHE.allow`optionally, if you want to persist allowance for those variables for later use.
```

    )
  );
  TFHE.allowThis(yBytes256);
  TFHE.allow(yBytes256, msg.sender);
}
```

## 2. Casting between encrypted types

This type of casting is used to reinterpret or convert one encrypted type into another. For example:

- `euint32` → `euint64`
- `ebytes128` → `ebytes256`

Casting between encrypted types is often required when working with operations that demand specific sizes or precisions.

> **Important**: When casting between encrypted types:
>
> - Casting from smaller types to larger types (e.g. `euint32` → `euint64`) preserves all information
> - Casting from larger types to smaller types (e.g. `euint64` → `euint32`) will truncate and lose information

The table below summarizes the available casting functions:

| From type | To type  | Function         |
| --------- | -------- | ---------------- |
| `euintX`  | `euintX` | `TFHE.asEuintXX` |
| `ebool`   | `euintX` | `TFHE.asEuintXX` |

> | `euintX`  | `ebool`  | `TFHE.asEboolXX` |

Reviewer: It is wrong. Please remove this line: casting from euintXX to ebool (contrarily to the opposite direction, ebool to euintX is good, keep it) is not supported and ill defined, since different languages do it in different manners. If you really want to convert a euintX to an ebool, you could use TFHE.Ne for instance. As suggested in TFHEExecutor comment already , see comment at the end of this line https://github.com/zama-ai/fhevm-backend/blob/main/contracts/contracts/TFHEExecutor.sol#L492

Author: ```
    // Cast an encrypted integer from euint8 to ebool.
    function asEbool(euint8 value) internal returns (ebool) {
        return ne(value, 0);
    }
```

Author: you have function definitions like so in the code

Reviewer: ah good catch, but as you see since this is using ne under the hood i forgot we abstracted away in TFHE. My bad, you can keep it, i had confusion because i thought about the TFHEExecutor implementation instead

{% hint style="info" %}
Casting between encrypted types is efficient and often necessary when handling data with differing precision requirements.
{% endhint %}

### **Workflow for encrypted types**

```solidity
// Casting between encrypted types
euint32 value32 = TFHE.asEuint32(value64); // Cast to euint32
ebool valueBool = TFHE.asEbool(value32);   // Cast to ebool
```

> ## 3. Encrypted input

Author: ```suggestion
## 3. Encrypted encrypt
```

### Overview

Encrypted input casting is the process of interpreting a handle (ciphertext reference) and its proof as a specific encrypted type. This ensures the validity of the input before it is used in computations.

Encrypted inputs is in depth explained in the following section: [encrypted inputs](./inputs.md)

#### Example

```solidity
euint64 encryptedValue = TFHE.asEuint64(einputHandle, inputProof); // Interpret einputHandle as euint64
```

#### Details

Encrypted input casting validates:

1.  The input handle references a valid ciphertext.
2.  The accompanying proof matches the expected type.

For more information, see the [Encrypetd inputs documentation](./inputs.md)

## Overall operation summary

| Casting Type             | Function                | Input Type        | Output Type |
| ------------------------ | ----------------------- | ----------------- | ----------- |

> | Trivial input            | `TFHE.asEuintXX(x)`     | `uintX`           | `euintX`    |

Reviewer: btw you forgot the eaddress type here, which is supported for trivial encryption and encrypted input (but not casting)

> | Trivial input            | `TFHE.asEuintXX(x)`     | `uintX`           | `euintX`    |

Reviewer: I would use `Trivial encryption` everywhere in place of `Trivial innput`

> | Trivial input            | `TFHE.asEuintXX(x)`     | `uintX`           | `euintX`    |

Author: ```suggestion
| Trivial encrypt            | `TFHE.asEuintXX(x)`     | `uintX`           | `euintX`    |
```

|                          | `TFHE.asEbool(x)`       | `bool`            | `ebool`     |
|                          | `TFHE.asEbytesXX(x)`    | `bytesXX`         | `ebytesXX`  |
| Conversion between types | `TFHE.asEuintXX(x)`     | `euintXX`/`ebool` | `euintYY`   |

> |                          | `TFHE.asEbool(x)`       | `euintXX`         | `ebool`     |

Reviewer: this TFHE.asEbool(x) with euintXX as input is wrong once again, and should be removed. We cannot cast to ebool

> | Encrypted input          | `TFHE.asEuintXX(x, y)`  | `einput`, proof   | `euintX`    |

Author: ```suggestion
| Encrypted input          | `TFHE.asEuintXX(x, y)`  | `einput`, `bytes` proof   | `euintX`    |
```

> |                          | `TFHE.asEbool(x, y)`    | `einput`,proof    | `ebool`     |

Reviewer: I would keep the `bytes` somewhere as second arguments for all those functions.  `einput, bytes proof` for eg

Author: ```suggestion
|                          | `TFHE.asEbool(x, y)`    | `einput`, `bytes` proof    | `ebool`     |
```

> |                          | `TFHE.asEbytesXX(x, y)` | `einput`,proof    | `ebytesXX`  |

Reviewer: last line also add the `bytes` before proof (please apply the change yourself, you can stop suggesting me many commits for each line, i might miss some and have to work on some other stuff)

> # asEbool, asEuintXX, asEaddress and asEbytesXX operations

Reviewer: I would have renamed this file `asEXXoperators.md`

> This documentation covers the `asEbool`, `asEuintXX`, and `asEbytesXX` operations provided by the TFHE library for working with encrypted data in the fhEVM. These operations are essential for converting between plaintext and encrypted types, as well as handling encrypted inputs.

Reviewer: also add mention of asEaddress here please

The operations can be categorized into three main use cases:

1. **Trivial encryption**: Converting plaintext values to encrypted types
2. **Type casting**: Converting between different encrypted types
3. **Input handling**: Processing encrypted inputs with proofs

## 1. Trivial encryption

Trivial encryption simply put is a plain text in a format of a ciphertext.

### Overview

Trivial encryption is the process of converting plaintext values into encrypted types (ciphertexts) compatible with TFHE operators. Although the data is in ciphertext format, it remains publicly visible on-chain, making it useful for operations between public and private values.

This type of casting involves converting plaintext (unencrypted) values into their encrypted equivalents, such as:

- `bool` → `ebool`
- `uint` → `euintXX`
- `bytes` → `ebytesXX`

> 

Reviewer: also add here address -> eaddress, this is possible for trivial encryption fyi

ebool valueBool = TFHE.asEbool(value32);   // Cast to ebool
```

## 3. Encrypted input

### Overview

Encrypted input casting is the process of interpreting a handle (ciphertext reference) and its proof as a specific encrypted type. This ensures the validity of the input before it is used in computations.

Encrypted inputs is in depth explained in the following section: [encrypted inputs](./inputs.md)

#### Example

```solidity
euint64 encryptedValue = TFHE.asEuint64(einputHandle, inputProof); // Interpret einputHandle as euint64
```

#### Details

Encrypted input casting validates:

1.  The input handle references a valid ciphertext.
2.  The accompanying proof matches the expected type.

For more information, see the [Encrypetd inputs documentation](./inputs.md)

## Overall operation summary

| Casting Type             | Function                | Input Type              | Output Type |
| ------------------------ | ----------------------- | ----------------------- | ----------- |
| Trivial encryption       | `TFHE.asEuintXX(x)`     | `uintX`                 | `euintX`    |
|                          | `TFHE.asEbool(x)`       | `bool`                  | `ebool`     |

> |                          | `TFHE.asEbytesXX(x)`    | `bytesXX`               | `ebytesXX`  |

Reviewer: same here, in trivial encryption part of the table, add eadress
